src/Core/MySQL/MySQLJsonConv.cpp

```diff
@@ -0,0 +1,940 @@
+                Int64 intpart = static_cast<Int64>(ultime >> 24);
+                Int32 frac = static_cast<Int32> (ultime % (1L << 24));
+                Int64 ymd = intpart >> 17;
+                Int64 ym = ymd >> 5;
+                Int64 hms = intpart % (1 << 17);
+                text.reserve(MAX_DATETIME_LENGTH + 1);
+                if (field_type == MYSQL_TYPE_DATE)
+                {
+                    text.resize(MAX_DATE_LENGTH);
+                    (void) sprintf(text.data(), "%04d-%02d-%02d",
+                        static_cast<Int32> (ym / 13),
+                        static_cast<Int32> (ym % 13),
+                        static_cast<Int32> (ymd % (1 << 5)));
+
+                }
+                else
+                {
+                    text.resize(MAX_DATETIME_LENGTH);
+                    (void) sprintf(text.data(), "%04d-%02d-%02d %02d:%02d:%02d.%06d",
+                        static_cast<Int32> (ym / 13),
+                        static_cast<Int32> (ym % 13),
+                        static_cast<Int32> (ymd % (1 << 5)),
+                        static_cast<Int32> (hms >> 12),
+                        static_cast<Int32> ((hms >> 6) % (1 << 6)),
+                        static_cast<Int32> (hms % (1 << 6)),
+                        frac);
+
+                }
+            }
+            out += ('"' + text +'"');
+            break;
+        }
+        default:
+        {
+            text = String(data, data_length);
+            out += '"';
+            escape(text, out);
+            out += '"';
+            break;
+        }
+    }
+}
+
+void JsonValue::toJsonString(String & out)
+{
+    switch (json_type)
+    {
+        case JsonEnumType::OBJECT:
+            out += '{';
+            for (size_t i = 0; i < element_count; i++)
+            {
+                if (i > 0)
+                {
+                    out += ", ";
+                }
+                out += '"';
+                String key = getObjectKey(i);
+                escape(key, out);
+                out += "\": ";
+                getJsonValueForElement(i).toJsonString(out);
+            }
+
+            out += '}';
+            break;
+
+        case JsonEnumType::ARRAY:
+            out += '[';
+            for (size_t i = 0; i < element_count; i++)
+            {
+                if (i > 0)
+                {
+                    out += ", ";
+                }
+                getJsonValueForElement(i).toJsonString(out);
+            }
+            out += ']';
+            break;
+
+        case JsonEnumType::DOUBLE:
+        {
+            // ostringstream will remove the tail '0' automatically
+            std::ostringstream oss; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+            // the default precision is 6, only 15 precision is valid. bigger precision is invalid data.
+            // e.g. 3.1415926535 with precision 32 will get 3.1415926535000000541231202078052
+            oss << std::setprecision(DOUBLE_PRECISION) << double_value;
```

Comment:
We can use double_conversion for a better algorithm, but see WriteHelpers where it is already done right.